Amplify/Categories/DataStore/Model/Temporal/TimeZone+Extension.swift

```diff
@@ -0,0 +1,108 @@
+            func hasMatch(regex: NSRegularExpression?, str: String) -> Bool {
+                return regex.flatMap {
+                    $0.firstMatch(in: str, range: NSRange(location: 0, length: str.count))
+                } != nil
+            }
+
+            // <time>±hh:mm
+            func suffixHHColonMM() -> String? {
+                if iso8601DateString.count > 6 {
+                    let tz = String(iso8601DateString.dropFirst(iso8601DateString.count - 6))
+                    if hasMatch(regex: TimeZone.iso8601TimeZoneHHColonMMRegex, str: tz) {
+                        return tz
+                    }
+                }
+                return nil
+            }
+
+            // <time>±hhmm
+            func suffixHHMM() -> String? {
+                if iso8601DateString.count > 5 {
+                    let tz = String(iso8601DateString.dropFirst(iso8601DateString.count - 5))
+                    if hasMatch(regex: TimeZone.iso8601TimeZoneHHMMRegex, str: tz) {
+                        return tz
+                    }
+                }
+                return nil
+            }
+
+            // <time>±hh
+            func suffixHH() -> String? {
+                if iso8601DateString.count > 3 {
+                    let tz = String(iso8601DateString.dropFirst(iso8601DateString.count - 3))
+                    if hasMatch(regex: TimeZone.iso8601TimeZoneHHRegex, str: tz) {
+                        return tz
+                    }
+                }
+                return nil
+            }
+
+            if iso8601DateString.hasPrefix("Z") { // <time>Z
+                self = .utc
+                return
+            }
+
+            if let tz = suffixHHColonMM(),
+               let hours = Int(tz.dropLast(3)),
+               let minutes = Int(tz.dropFirst(4))
+            {
+                self = .hhmm(hours: hours, minutes: minutes)
+                return
+            }
+
+            if let tz = suffixHHMM(),
+               let hours = Int(tz.dropLast(2)),
+               let minutes = Int(tz.dropFirst(3))
+            {
+                self = .hhmm(hours: hours, minutes: minutes)
+                return
+            }
+
+            if let tz = suffixHH(),
+               let hours = Int(tz)
+            {
+                self = .hh(hours: hours)
+                return
+            }
+
+            return nil
+        }
+    }
+
+    /// https://en.wikipedia.org/wiki/ISO_8601#Time_zone_designators
+    @usableFromInline
+    internal init?(iso8601DateString: String) {
+        switch ISO8601TimeZonePart(iso8601DateString: iso8601DateString) {
+        case .some(.utc):
+            self.init(abbreviation: "UTC")
+        case let .some(.hh(hours: hours)):
+            self.init(secondsFromGMT: hours * 60 * 60)
+        case let .some(.hhmm(hours: hours, minutes: minutes)):
+            self.init(secondsFromGMT: hours * 60 * 60 +
+                      (hours > 0 ? 1 : -1) * minutes * 60)
```

Comment:
Should we support working with AppSync's AWSDateTime?

In https://docs.aws.amazon.com/appsync/latest/devguide/scalars.html AppSync supports 

> ±hh:mm:ss. The seconds field in the time zone offset is considered valid even though it's not part of the ISO 8601 standard.

Currently the parsing here does not support seconds in the offset. Regardless of the decision to support it now with an iteration on this PR, i think we should still add unit tests in ~~DataStore and API~~ AWSPluginsCore to assert for the expected behavior when decoding datetimes with timezone offsets containing seconds.

The expected behavior without any changes in this PR would be that the parser works and drops the seconds. If we decide to support seconds, then we should keep the unit tests with seconds in the Amplify unit tests, and call out in the implementation that it is an extension of the ISO standard

Reply:
Good catch, I didn't know AWSDateTime supports that format. I will add support for `±hh:mm:ss`